x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplicationIndexService.java

```diff
@@ -488,5 +462,31 @@ static void writeSearchApplicationBinaryWithVersion(SearchApplication app, Outpu
         }
     }
 
+    static class DelegatingIndexNotFoundActionListener<T, R> extends ActionListener.Delegating<T, R> {
+
+        private final BiConsumer<ActionListener<R>, T> bc;
+        private final String resourceName;
+
+        DelegatingIndexNotFoundActionListener(String resourceName, ActionListener<R> delegate, BiConsumer<ActionListener<R>, T> bc) {
+            super(delegate);
+            this.bc = bc;
+            this.resourceName = resourceName;
+        }
+
+        @Override
+        public void onResponse(T t) {
+            bc.accept(delegate, t);
+        }
+
+        @Override
+        public void onFailure(Exception e) {
+            if (e instanceof IndexNotFoundException) {
+                delegate.onFailure(new ResourceNotFoundException(resourceName));
```

Comment:
Might be useful to keep hold of the original stack trace here too - something like this should do the trick I think:

```suggestion
                delegate.onFailure(new ResourceNotFoundException(resourceName, e));
```